src/metadata_value.rs

```diff
@@ -14,39 +20,33 @@ use metadata_key::MetadataKey;
 /// [`HeaderMap`]: struct.HeaderMap.html
 #[derive(Clone, Hash)]
 #[repr(transparent)]
-pub struct MetadataValue {
+pub struct MetadataValue<VE: ValueEncoding> {
     // Note: There are unsafe transmutes that assume that the memory layout
     // of MetadataValue is identical to HeaderValue
     pub(crate) inner: HeaderValue,
+    phantom: PhantomData<VE>,
 }
 
-/// A possible error when converting a `MetadataValue` from a string or byte
-/// slice.
-#[derive(Debug)]
-pub struct InvalidMetadataValue {
-    _priv: (),
-}
-
-/// A possible error when converting a `MetadataValue` from a string or byte
-/// slice.
-#[derive(Debug)]
-pub struct InvalidMetadataValueBytes(InvalidMetadataValue);
-
 /// A possible error when converting a `MetadataValue` to a string representation.
 ///
 /// Metadata field values may contain opaque bytes, in which case it is not
 /// possible to represent the value as a string.
-#[derive(Debug)]
+#[derive(Debug, Default)]
```

Comment:
I'd recommend removing the `Default`, since it a) allows creating a `ToStrError` outside the crate, and b) can possibly interfere if needing to add a field later that does not implement `Default`.

Reply:
(This comment is a response to all of your comments in this review related to this.)

I'm open to changing my mind here, especially since I'm a Rust newbie, but to me hiding constructors of error types seems like a bad idea. I can't recall seeing a design in some other language that attempts to do this.

Conceptually, I think of error and exception types as value types (as in [value vs identity](https://www.youtube.com/watch?v=-6BsiVyC1kM)), and, being value types, I do not see any issue in allowing anyone to create any kind of error value they desire.

Practically I care about this because of testing. Disallowing any other code to create these error values requires any code that uses this to immediately convert the error to some other type, even thin wrappers that don't do much, otherwise it is impossible to test error conditions.

It's true that it becomes impossible to add mandatory non-defaultable fields without an API break but the probability of that happening, combined with the small pain of being forced to use `Option` or similar, vs the relatively large testability pain makes it seem like a bad tradeoff to not allow tests to create empty error values.

Reply:
My $0.02 is to agree with @seanmonstar on this. Although I definitely understand @per-gron's point here, in my experience, it's pretty important for libraries to avoid anything that has the potential to cause a breaking change, even if it doesn't seem particularly likely. And, as @seanmonstar points out, this error is relatively easy to generate in testing situations.

Reply:
Thanks for taking the time to explain more thoroughly. It still seems a bit odd to me (especially given that the worst case scenario is to have to use `Option`) but I see your point about being conservative. I've updated the PR to make the constructors `pub(crate)` and remove `Default`.